packages/astro/src/env/schema.ts

```diff
@@ -75,11 +75,28 @@ const SecretServerEnvFieldMetadata = z.object({
 	context: z.literal('server'),
 	access: z.literal('secret'),
 });
-const EnvFieldMetadata = z.union([
+const _EnvFieldMetadata = z.union([
 	PublicClientEnvFieldMetadata,
 	PublicServerEnvFieldMetadata,
 	SecretServerEnvFieldMetadata,
 ]);
+const EnvFieldMetadata = z.custom<z.input<typeof _EnvFieldMetadata>>().superRefine((data, ctx) => {
+	const result = _EnvFieldMetadata.safeParse(data);
+	if (result.success) {
+		return;
+	}
```

Comment:
Can you explain to me where we check the **values** of `access` and `context`? I can't see it in the code, it's probably somewhere

Reply:
The `_EnvFieldMetadata` schema (formerly `EnvFieldMetadata`) is a union of all possibilities (3). It takes care of validating the values (so here the validation happens during `_EnvFieldMetadata.safeParse()`) but the message was a bit cryptic as reported by Erika, so I'm wrapping it with a custom schema. Hope that helps!

Reply:
Alright so I've tried a few things. Taking this invalid combination:
```js
            FOO: envField.string({
                context: "client",
                access: "secret",
            })
```
throws
```
! env.schema.FOO: Did not match union.
  > Expected type { access: public } | { context: server; access: public } | { context: server }
  > Received { "context": "client", "access": "secret", "type": "string" }
```
I tried setting `message` on every metadata schema, `errorMap` as well but it's not surfaced. I think what's in the PR is the best way to have a proper custom message

Reply:
That's a pity :( thank you for checking